content/blog/2023-12-11-jitnic/index.md

```diff
@@ -0,0 +1,129 @@
+
+For this project, we focus on the compiler and implement the “hardware” NIC as an eBPF simulator. For this reason, we do not go into detail about the hardware implementation of certain features. Our system is built in ~1400 of Ocaml code, along with some tools for parsing out the eBPF byte code of an object file and turning PCAP (packet capture) files into byte streams. There are also about ~450 lines of an experimental Rust implementation that is now defunct.
+
+We are actively working on different memory consistency models – sharing memory between a smartNIC and CPU gets very complicated, but for the sake of the system we implemented for this class, we assume the programs are stateless and there are no memory reads/writes outside of the input packet data. This also removes the problem of having memory reads in path conditions for memory locations that are also written to (you can quickly see how complicated this would get…). 
+
+## Case study: Ipv6 Filter
+
+The following C code is an XDP program that filters all non-ipv6 traffic. This will call the XDP_DROP action on any packets that do not match the ipv6 protocol in their ethernet header. Any ipv6 packets, or malformed packets that do not pass the bounds check will be passed onto the linux kernel networking stack. 
+```C
+int xdp_ipv6_filter_program(struct xdp_md *ctx) 
+{ 
+void *data_end = ctx->data_end; /* pointer to where the packet ends in memory */
+struct ethhdr *eth = ctx->data; /* packet as an ethernet struct */
+if ((void *) eth + sizeof(*eth) > data_end) /* out of bounds check for packet read */
+	return XDP_PASS;
+
+if (eth->protocol == 0x86dd) /* bytes of eth type ipv6 */
+	return XDP_PASS;
+else 
+	return XDP_DROP;
+}  
+```
+
+For the sake of simplicity, and to avoid introducing eBPF assembly, we will keep this example at the level of C. Note every C instruction has an analogous few instructions in eBPF. 
+
+Starting from the beginning, let’s assume a packet arrives that has an ipv4 ethernet header. Taking the execution model from the diagram, this will be a cold miss as the cache is empty. The JIT compiler on the CPU will trace the program with the packet as input and produce the following trace:
+```C
+void *data_end = ctx->data_end; /* pointer to where the packet ends in memory */
+struct ethhdr *eth = ctx->data;
+return XDP_DROP; 
+```
+The path condition will be `eth + sizeof(*eth) <= data_end && eth->protocol != 0x86dd` – notice that when tracing conditions just become part of the path condition. (Also note: in C this is `eth->protocol` but in eBPF this becomes a lookup to an offset within the packet. Similarly sizeof(*eth) is a constant and eth and data_end are given as input). At this point, the CPU will drop the packet. 
```

Comment:
One important thing I couldn't quite understand from this example, which I think you are getting at partially here: this path condition expression, as written, contains variables that are defined in the trace. To deal with this in general, I can think of a couple of options:

* evaluate the path condition after running the trace, but before making a drop/forward decision
* "slice" the computation of all of those variables, so the path condition is a self-contained thing (at the cost of possibly recomputing some values when we do eventually run the trace)
* somehow interleaving the trace and condition evaluation so we can make a decision as soon as all the "ingredients" for the path condition are available

Maybe it would be good to be explicit about exactly what the current system does?

Reply:
We have clarified this in the writeup. In summary, we currently produce path conditions that are evaluated from offsets within the packet - in our system we just store memory offset values and then calculate the path condition based on those. If a path condition depends some address that isn't a constant offset, we abort tracing and run the concrete interpreter to execute the program on the packet. We discuss a possible approach for producing traces of this type in the post.

Reply:
Got it; that makes sense!!

Reply:
To clarify, we *do* support non-constant offsets; we abort tracing only when a branch depends on the value stored in a memory location that has already been written to.